.github/workflows/release.yaml

```
@@ -0,0 +1,118 @@
+name: build and release
+
+on:
+  push:
+    tags:
+      - 'v*'
+
+env:
+  # https://github.com/llvm/llvm-project/releases/tag/llvmorg-15.0.0
+  LLVM_arm64: clang+llvm-15.0.0-aarch64-linux-gnu
+  LLVM_amd64: clang+llvm-15.0.0-x86_64-linux-gnu-rhel-8.4 
+
+jobs:
+  build:
+    name: Build static bpftool binary
+    runs-on: ubuntu-22.04
+    env:
+      TARGETARCH: ${{ matrix.arch }}
+      FILE_STRING_ARCH_amd64: x86-64
+      FILE_STRING_ARCH_arm64: aarch64
+    strategy:
+      matrix:
+        arch: [arm64, amd64]
+
+    steps:
+      # amd64 needs the dependencies to build bpftool
+      - name: Install dependencies (amd64)
+        if: matrix.arch == 'amd64'
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y libelf-dev libcap-dev llvm
+
+      - name: Download and extract compiled LLVM release
+        env:
+          LLVM: ${{ env[format('LLVM_{0}', matrix.arch)] }}
+        run: |
+          curl -L https://github.com/llvm/llvm-project/releases/download/llvmorg-15.0.0/${{ env.LLVM }}.tar.xz -O
+          tar xvf ${{ env.LLVM }}.tar.xz
+
+      - name: Checkout bpftool code
+        uses: actions/checkout@24cb9080177205b6e8c946b17badbe402adc938f # v3.4.0
+        with:
+          submodules: recursive
+          path: 'bpftool'
+
+      - name: Build static bpftool natively for amd64
+        if: matrix.arch == 'amd64'
+        working-directory: 'bpftool'
+        env:
+          LLVM: ${{ env[format('LLVM_{0}', matrix.arch)] }}
+        run: make -j $(nproc) -C src V=1 \ 
+          EXTRA_CFLAGS=--static \
+          LLVM_CONFIG="${GITHUB_WORKSPACE}/${{ env.LLVM }}/bin/llvm-config" \
+          LLVM_STRIP="${GITHUB_WORKSPACE}/${{ env.LLVM }}/bin/llvm-strip" &&
+          /usr/bin/strip src/bpftool
+
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@e81a89b1732b9c48d79cd809d8d81d79c4647a18 # v2.1.0
+        if: matrix.arch == 'arm64'
+        with:
+          platforms: arm64
+
+      # The emulated build leverages Docker and Ubuntu 22.04 container image
+      # distribution to have all the needed arm64 packages.
+      - name: Build static bpftool for arm64 with emulation
+        if: matrix.arch == 'arm64'
+        env:
+          LLVM: ${{ env[format('LLVM_{0}', matrix.arch)] }}
+        run:  |
+          docker run --platform linux/arm64 --rm -v $(pwd):/build ubuntu:22.04 \
+          bash -c "apt-get update && \
+                   apt-get install -y make pkg-config libelf-dev libcap-dev clang gcc llvm binutils && \
+                   cd build/bpftool && \
+                   LLVM_CONFIG=../../${{ env.LLVM }}/bin/llvm-config EXTRA_CFLAGS=--static make -j V=1 -C src && \
+                   strip src/bpftool"
+
+      - name: Test bpftool binary
+        working-directory: 'bpftool/src'
+        env:
+          ARCH: ${{ env[format('FILE_STRING_ARCH_{0}', matrix.arch)] }}
+        run: |
+          file ./bpftool | grep ${{ env.ARCH }}
```

Thanks! Can we please have like for the other tests, `tee` to `stderr` to keep the output, and `grep -q` to produce an error if the pattern is missing?

```suggestion
          file ./bpftool | \
              tee /dev/stderr | \
              grep -q "${{ env.ARCH }}"
```

---

To me these are equivalents, you just duplicate the output before suppressing it.

---

but if you prefer stderr I am okay :D I just added your suggestion

---

It's not that I prefer `stderr`, but I prefer to see the output of the command, too. You don't get the output from `file` in the workflow logs if you pipe it to `grep` without duplicating it, do you?

---

you do! https://github.com/mtardy/bpftool/actions/runs/4477949024/jobs/7870109343#step:8:19 it outputs the lines that matched.

---

but it doesn't really matter, it's the same

---

Ah of course, it's a single-line output :facepalm: OK feel free to use one or the other